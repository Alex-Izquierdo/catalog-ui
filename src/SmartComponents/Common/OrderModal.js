import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Modal } from '@patternfly/react-core';

import CatItemSvg from '../../assets/images/vendor-openshift.svg';
import ImageWithDefault from '../../presentational-components/shared/image-with-default';
import OrderServiceFormStepConfiguration from '../Order/OrderServiceFormStepConfiguration';

const OrderModal = ({ serviceData, closeUrl, history: { push }}) => serviceData ? (
  <Modal
    isOpen
    hideTitle
    onClose={ () => push(closeUrl) }
    style={ { maxWidth: 800, minHeight: 300 } }
  >
    <ImageWithDefault src = { serviceData.imageUrl || CatItemSvg } width="40" />
    { serviceData.name }
    <OrderServiceFormStepConfiguration closeUrl={ closeUrl } { ...serviceData } />
  </Modal>
) : null;

OrderModal.propTypes = {
  orderData: PropTypes.func,
  serviceData: PropTypes.object,
  closeUrl: PropTypes.string.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};

const mapStateToProps = ({
  orderReducer: { selectedItem, servicePlans }
}) => ({
  selectedItem,
  servicePlans
});

export default withRouter(connect(mapStateToProps)(OrderModal));
