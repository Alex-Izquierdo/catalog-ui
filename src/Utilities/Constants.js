export const DEFAULT_PAGE_SIZE = 50;

const topoApiHost = process.env.TOPO_API_HOST || '';
const topoApiPort = process.env.TOPO_API_PORT || '443';
const topoBasePath = process.env.TOPO_BASE_PATH;
const topoApiVersion = process.env.TOPO_API_VERSION || '/api/v0.0';
export const SOURCES_API_BASE = `https://${topoApiHost}:${topoApiPort}${topoBasePath}${topoApiVersion.replace(/\/+$/, '')}`;

const apiHost = process.env.API_HOST || 'localhost';
const apiPort = process.env.API_PORT || '5000';
const basePath = process.env.BASE_PATH || '';
const apiVersion = process.env.API_VERSION || '/api/v0.0';
export const SERVICE_PORTAL_API_BASE = `https://${apiHost}:${apiPort}${basePath}${apiVersion.replace(/\/+$/, '')}`;
