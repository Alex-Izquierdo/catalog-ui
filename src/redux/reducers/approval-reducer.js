import { SET_LOADING_STATE } from '../action-types';
import { ASYNC_ACTIONS } from '../action-types/approval-action-types';

export const approvalInitialState = {
  isFetching: false,
  isResolving: false,
  workflows: [],
  resolvedWorkflows: []
};

const setLoadingState = (state, { payload = true }) => ({
  ...state,
  isFetching: payload
});
const setWorkflows = (state, { payload }) => ({
  ...state,
  isFetching: false,
  workflows: payload
});
const setResolvingState = (state, { payload = true }) => ({
  ...state,
  isResolving: payload
});
const setResolvedWorkflows = (state, { payload }) => ({
  ...state,
  isResolving: false,
  resolvedWorkflows: payload
});

export default {
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_PENDING]: setLoadingState,
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_FULFILLED]: setWorkflows,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_PENDING]: setResolvingState,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_FULFILLED]: setResolvedWorkflows,
  [SET_LOADING_STATE]: setLoadingState
};
