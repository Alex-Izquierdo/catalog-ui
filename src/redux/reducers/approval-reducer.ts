import { Workflow } from '@redhat-cloud-services/approval-client';
import { SET_LOADING_STATE } from '../action-types';
import { ASYNC_ACTIONS } from '../action-types/approval-action-types';
import { StateFromReducersMapObject } from 'redux';
import { ReduxAction, AnyObject } from '../../types/common-types';

export interface ApprovalReducerState extends AnyObject {
  isFetching: boolean;
  isResolving: boolean;
  workflows: Workflow[];
  resolvedWorkflows: Workflow[];
}

export type ApprovalReducerActionHandler = (
  state: StateFromReducersMapObject<ApprovalReducerState>,
  action: ReduxAction
) => ApprovalReducerState;

export const approvalInitialState = {
  isFetching: false,
  isResolving: false,
  workflows: [],
  resolvedWorkflows: []
};

const setLoadingState: ApprovalReducerActionHandler = (
  state,
  { payload = true }
) => ({
  ...state,
  isFetching: payload
});
const setWorkflows: ApprovalReducerActionHandler = (state, { payload }) => ({
  ...state,
  isFetching: false,
  workflows: payload
});
const setResolvingState: ApprovalReducerActionHandler = (
  state,
  { payload = true }
) => ({
  ...state,
  isResolving: payload
});
const setResolvedWorkflows: ApprovalReducerActionHandler = (
  state,
  { payload }
) => ({
  ...state,
  isResolving: false,
  resolvedWorkflows: payload
});

export default {
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_PENDING]: setLoadingState,
  [ASYNC_ACTIONS.FETCH_WORKFLOWS_FULFILLED]: setWorkflows,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_PENDING]: setResolvingState,
  [ASYNC_ACTIONS.RESOLVE_WORKFLOWS_FULFILLED]: setResolvedWorkflows,
  [SET_LOADING_STATE]: setLoadingState
};
