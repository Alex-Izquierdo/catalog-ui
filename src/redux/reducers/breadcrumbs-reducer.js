import { INITIALIZE_BREADCRUMBS } from '../action-types';

export const initialBreadcrumbsState = {
  fragments: []
};

const initialize = (state, { payload }) => ({ ...state, fragments: payload });

export default {
  [INITIALIZE_BREADCRUMBS]: initialize
};
