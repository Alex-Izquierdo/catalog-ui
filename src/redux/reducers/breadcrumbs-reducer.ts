import { ReactNode } from 'react';
import { StateFromReducersMapObject } from 'redux';
import { INITIALIZE_BREADCRUMBS } from '../action-types';
import { AnyObject, ReduxAction } from '../../types/common-types';

export interface BreadcrumbFragment {
  pathname: string;
  searchParams?: string;
  title?: ReactNode;
}
export interface BreadcrumbsReducerState extends AnyObject {
  fragments: BreadcrumbFragment[];
}

export type BreadcrumbsReducerActionHandler = (
  state: StateFromReducersMapObject<BreadcrumbsReducerState>,
  action: ReduxAction
) => BreadcrumbsReducerState;

export const initialBreadcrumbsState: BreadcrumbsReducerState = {
  fragments: []
};

const initialize: BreadcrumbsReducerActionHandler = (state, { payload }) => ({
  ...state,
  fragments: payload
});

export default {
  [INITIALIZE_BREADCRUMBS]: initialize
};
