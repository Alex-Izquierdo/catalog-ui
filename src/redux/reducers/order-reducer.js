import {
  FETCH_SERVICE_PLANS,
  LIST_ORDERS,
  FETCH_SERVICE_PLAN_PARAMETERS,
  SUBMIT_SERVICE_ORDER,
  UPDATE_SERVICE_DATA,
  SET_SELECTED_PLAN,
  FETCH_REQUESTS,
  FETCH_ORDER_ITEMS,
  SET_LOADING_STATE,
  SET_ORDERS,
  FETCH_ORDERS,
  SET_ORDER_DETAIL
} from '../action-types';
import { defaultSettings } from '../../helpers/shared/pagination';
// Initial State
export const orderInitialState = {
  servicePlans: [],
  selectedPlan: {},
  serviceData: {},
  isLoading: false,
  requests: [],
  orderDetail: {
    order: {},
    portfolioItem: {},
    platform: {}
  },
  orders: {
    data: [],
    meta: { ...defaultSettings }
  }
};

const setLoadingState = (state, { payload = true }) => ({
  ...state,
  isLoading: payload
});
const setServicePlans = (state, { payload }) => ({
  ...state,
  servicePlans: payload,
  isLoading: false
});
const setListOrder = (state, { payload }) => ({
  ...state,
  orders: payload,
  isLoading: false
});
const setPlanParameters = (state, { payload }) => ({
  ...state,
  planParameters: payload,
  isLoading: false
});
const submitServiceOrder = (state, { payload }) => ({
  ...state,
  ...payload,
  isLoading: false
});
const updateServiceData = (state, { payload }) => ({
  ...state,
  serviceData: payload,
  isLoading: false
});
const selectPlan = (state, { payload }) => ({
  ...state,
  selectedPlan: payload,
  isLoading: false
});
const setRequests = (state, { payload }) => ({
  ...state,
  requests: payload,
  isLoading: false
});
const setOrderItems = (state, { payload }) => ({
  ...state,
  orderItems: payload,
  isLoading: false
});
const setOrders = (state, { payload }) => ({ ...state, orders: payload });
const setOrderDetail = (state, { payload }) => ({
  ...state,
  orderDetail: payload
});

export default {
  [`${FETCH_SERVICE_PLANS}_PENDING`]: setLoadingState,
  [`${FETCH_SERVICE_PLANS}_FULFILLED`]: setServicePlans,
  [`${LIST_ORDERS}_PENDING`]: setLoadingState,
  [`${LIST_ORDERS}_FULFILLED`]: setListOrder,
  [`${FETCH_SERVICE_PLAN_PARAMETERS}_PENDING`]: setLoadingState,
  [`${FETCH_SERVICE_PLAN_PARAMETERS}_FULFILLED`]: setPlanParameters,
  [`${SUBMIT_SERVICE_ORDER}_PENDING`]: setLoadingState,
  [`${SUBMIT_SERVICE_ORDER}_FULFILLED`]: submitServiceOrder,
  [UPDATE_SERVICE_DATA]: updateServiceData,
  [SET_SELECTED_PLAN]: selectPlan,
  [`${FETCH_REQUESTS}_PENDING`]: setLoadingState,
  [`${FETCH_REQUESTS}_FULFILLED`]: setRequests,
  [`${FETCH_ORDER_ITEMS}_PENDING`]: setLoadingState,
  [`${FETCH_ORDER_ITEMS}_FULFILLED`]: setOrderItems,
  [SET_LOADING_STATE]: setLoadingState,
  [`${FETCH_ORDERS}_FULFILLED`]: setOrders,
  [`${FETCH_ORDERS}_PENDING`]: setLoadingState,
  [SET_ORDERS]: setOrders,
  [`${SET_ORDER_DETAIL}_FULFILLED`]: setOrderDetail,
  [SET_ORDER_DETAIL]: setOrderDetail
};
