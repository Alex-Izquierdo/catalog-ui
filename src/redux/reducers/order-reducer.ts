import { StateFromReducersMapObject } from 'redux';
import { Request } from '@redhat-cloud-services/approval-client';
import {
  ServicePlan,
  Order,
  PortfolioItem,
  Portfolio
} from '@redhat-cloud-services/catalog-client';
import { Source } from '@redhat-cloud-services/sources-client';
import {
  FETCH_SERVICE_PLANS,
  LIST_ORDERS,
  FETCH_SERVICE_PLAN_PARAMETERS,
  SUBMIT_SERVICE_ORDER,
  SET_SELECTED_PLAN,
  FETCH_REQUESTS,
  FETCH_ORDER_ITEMS,
  SET_LOADING_STATE,
  SET_ORDERS,
  FETCH_ORDERS,
  SET_ORDER_DETAIL,
  FETCH_APPROVAL_REQUESTS
} from '../action-types';
import { defaultSettings } from '../../helpers/shared/pagination';
import {
  ApiCollectionResponse,
  ReduxAction,
  AnyObject
} from '../../types/common-types';

export interface OrderDetail extends AnyObject {
  order: Order;
  portfolioItem: PortfolioItem;
  platform: Source;
  portfolio: Portfolio;
}
export interface OrderReducerState extends AnyObject {
  servicePlans: ServicePlan[];
  selectedPlan: ServicePlan;
  isLoading: boolean;
  requests: Request[];
  orderDetail: OrderDetail;
  orders: ApiCollectionResponse<Order>;
}
// Initial State
export const orderInitialState: OrderReducerState = {
  servicePlans: [],
  selectedPlan: {},
  isLoading: false,
  requests: [],
  orderDetail: {
    order: {},
    portfolioItem: {},
    platform: {},
    portfolio: {}
  },
  orders: {
    data: [],
    meta: { ...defaultSettings }
  }
};

export type OrderReducerActionHandler = (
  state: StateFromReducersMapObject<OrderReducerState>,
  action: ReduxAction
) => OrderReducerState;

const setLoadingState: OrderReducerActionHandler = (
  state,
  { payload = true }
) => ({
  ...state,
  isLoading: payload
});
const setServicePlans: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  servicePlans: payload,
  isLoading: false
});
const setListOrder: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  orders: payload,
  isLoading: false
});
const setPlanParameters: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  planParameters: payload,
  isLoading: false
});
const submitServiceOrder: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  ...payload,
  isLoading: false
});
const selectPlan: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  selectedPlan: payload,
  isLoading: false
});
const setRequests: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  requests: payload,
  isLoading: false
});
const setOrderItems: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  orderItems: payload,
  isLoading: false
});
const setOrders: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  orders: payload
});
const setOrderDetail: OrderReducerActionHandler = (state, { payload }) => ({
  ...state,
  orderDetail: payload
});
const updateOrderApprovalRequests: OrderReducerActionHandler = (
  state,
  { payload }
) => ({
  ...state,
  orderDetail: {
    ...(state.orderDetail as OrderDetail),
    approvalRequest: payload
  }
});

export default {
  [`${FETCH_SERVICE_PLANS}_PENDING`]: setLoadingState,
  [`${FETCH_SERVICE_PLANS}_FULFILLED`]: setServicePlans,
  [`${LIST_ORDERS}_PENDING`]: setLoadingState,
  [`${LIST_ORDERS}_FULFILLED`]: setListOrder,
  [`${FETCH_SERVICE_PLAN_PARAMETERS}_PENDING`]: setLoadingState,
  [`${FETCH_SERVICE_PLAN_PARAMETERS}_FULFILLED`]: setPlanParameters,
  [`${SUBMIT_SERVICE_ORDER}_PENDING`]: setLoadingState,
  [`${SUBMIT_SERVICE_ORDER}_FULFILLED`]: submitServiceOrder,
  [SET_SELECTED_PLAN]: selectPlan,
  [`${FETCH_REQUESTS}_PENDING`]: setLoadingState,
  [`${FETCH_REQUESTS}_FULFILLED`]: setRequests,
  [`${FETCH_ORDER_ITEMS}_PENDING`]: setLoadingState,
  [`${FETCH_ORDER_ITEMS}_FULFILLED`]: setOrderItems,
  [SET_LOADING_STATE]: setLoadingState,
  [`${FETCH_ORDERS}_FULFILLED`]: setOrders,
  [`${FETCH_ORDERS}_PENDING`]: setLoadingState,
  [SET_ORDERS]: setOrders,
  [`${SET_ORDER_DETAIL}_FULFILLED`]: setOrderDetail,
  [SET_ORDER_DETAIL]: setOrderDetail,
  [`${FETCH_APPROVAL_REQUESTS}_FULFILLED`]: updateOrderApprovalRequests
};
