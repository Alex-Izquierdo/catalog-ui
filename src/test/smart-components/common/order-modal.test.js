import React from 'react';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store' ;
import { shallowToJson } from 'enzyme-to-json';
import { Button } from '@patternfly/react-core';
import { MemoryRouter, Route } from 'react-router-dom';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@red-hat-insights/insights-frontend-components/components/Notifications';

import dummySchema from '../order/order-mock-form-schema';
import { CATALOG_API_BASE } from '../../../Utilities/Constants';
import OrderModal from '../../../smart-components/common/order-modal';
import { orderInitialState } from '../../../redux/reducers/order-reducer';

describe('<OrderModal />', () => {
  let initialProps;
  const middlewares = [ thunk, promiseMiddleware(), notificationsMiddleware() ];
  let mockStore;
  let initialState;

  const OrderWrapper = ({ store, children, initialEntries = []}) => (
    <Provider store={ store }>
      <MemoryRouter initialEntries={ initialEntries }>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      closeUrl: '/close',
      serviceData: {
        name: 'Foo',
        id: '1'
      }
    };
    mockStore = configureStore(middlewares);
    initialState = {
      orderReducer: {
        ...orderInitialState,
        serviceData: {
          name: 'Foo',
          id: '1'
        }
      }
    };
  });

  it('should render correctly', () => {
    const wrapper = shallow(<OrderModal { ...initialProps } />);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should redirect back to close URL', (done) => {
    const store = mockStore(initialState);

    apiClientMock.get(`${CATALOG_API_BASE}/portfolio_items/1/service_plans`, mockOnce({
      body: [ dummySchema ]
    }));

    fetchMock.getOnce(`${CATALOG_API_BASE}/portfolio_items/1/provider_control_parameters`, {
      required: [ 'namespace' ],
      type: 'object',
      properties: {
        namespace: {
          title: 'Namespace',
          enum: [ '1', '2', '3', '4' ]
        }
      }
    });

    const wrapper = mount(
      <OrderWrapper store={ store } initialEntries={ [ '/foo/url' ] }>
        <Route to="/foo/url" render={ args => <OrderModal { ...initialProps } { ...args } /> }  />
      </OrderWrapper>
    );

    setImmediate(() => {
      wrapper.find(Button).first().simulate('click');
      expect(wrapper.find(MemoryRouter).instance().history.location.pathname).toEqual('/close');
      done();
    });
  });
});

