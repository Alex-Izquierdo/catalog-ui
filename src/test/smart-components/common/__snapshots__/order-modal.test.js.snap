// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderModal /> should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={Array []}
  >
    <OrderModal
      closeUrl="/close"
      serviceData={
        Object {
          "id": "1",
          "name": "Foo",
        }
      }
    />
  </MemoryRouter>
</Provider>
`;
